torch/lib/c10d/TcpStore.cpp

```diff
@@ -0,0 +1,326 @@
+         * exception, other connections will get an exception once they try to
+         * use the store. We will go ahead and close this connection whenever
+         * we hit an exception here.
+         */
+        ::close(fds[fdIdx].fd);
+        fds.erase(fds.begin() + fdIdx);
+        sockets_.erase(sockets_.begin() + fdIdx - 2);
+        --fdIdx;
+        continue;
+      }
+    }
+  }
+}
+
+void TCPStoreDaemon::stop() {
+  if (controlPipeFd_.size() == 2 && controlPipeFd_[1] != -1) {
+    // close the write end of the pipe
+    ::close(controlPipeFd_[1]);
+    controlPipeFd_[1] = -1;
+  }
+}
+
+/**
+ * query communicates with the worker. The format
+ * of the query is as follows:
+ * type of query | size of arg1 | arg1 | size of arg2 | arg2 | ...
+ * or, in the case of wait
+ * type of query | number of args | size of arg1 | arg1 | ...
+ */
+void TCPStoreDaemon::query(int socket) {
+
+  QueryType qt;
+  tcputil::recvBytes<QueryType>(socket, &qt, 1);
+
+  if (qt == QueryType::SET) {
+    setHandler(socket);
+
+  } else if (qt == QueryType::ADD) {
+    addHandler(socket);
+
+  } else if (qt == QueryType::GET) {
+    getHandler(socket);
+
+  } else if (qt == QueryType::CHECK) {
+    checkHandler(socket);
+
+  } else {
+    throw std::runtime_error("Unexpected query type");
+  }
+}
+
+void TCPStoreDaemon::setHandler(int socket) {
+  std::string key = tcputil::recvString(socket);
+  tcpStore_[key] = tcputil::recvVector<uint8_t>(socket);
+}
+
+void TCPStoreDaemon::addHandler(int socket) {
+  std::string key = tcputil::recvString(socket);
+  int64_t addVal = tcputil::recvValue<int64_t>(socket);
+
+  if (tcpStore_.find(key) != tcpStore_.end()) {
+    auto buf = reinterpret_cast<const char*>(tcpStore_[key].data());
+    auto len = tcpStore_[key].size();
+    addVal += std::stoll(std::string(buf, len));
+  }
+  auto addValStr = std::to_string(addVal);
+  tcpStore_[key] = std::vector<uint8_t>(addValStr.begin(), addValStr.end());
+  // Now send the new value
+  tcputil::sendValue<int64_t>(socket, addVal);
+}
+
+void TCPStoreDaemon::getHandler(int socket) {
+  std::string key = tcputil::recvString(socket);
+  auto data = tcpStore_.at(key);
+  tcputil::sendVector<uint8_t>(socket, data);
+}
+
+void TCPStoreDaemon::checkHandler(int socket) {
+  SizeType nargs;
+  tcputil::recvBytes<SizeType>(socket, &nargs, 1);
+  std::vector<std::string> keys(nargs);
+  for (size_t i = 0; i < nargs; i++) {
+    keys[i] = tcputil::recvString(socket);
+  }
+  // Now we have received all the keys
+  if (checkAndUpdate(keys)) {
+    tcputil::sendValue<CheckResponseType>(socket, CheckResponseType::READY);
+  } else {
+    tcputil::sendValue<CheckResponseType>(socket, CheckResponseType::NOT_READY);
+  }
+}
+
+bool TCPStoreDaemon::checkAndUpdate(std::vector<std::string>& keys) const {
+  bool ret = true;
+  for (auto it = keys.begin(); it != keys.end();) {
+    if (tcpStore_.count(*it) == 0) {
+      ret = false;
+      it++;
+    } else {
+      it = keys.erase(it);
```

Comment:
It looks like this code codes a lot of unnecessary things. I think the whole problem could be solved as (without modifying the vector):
```cpp 
return std::all_of(keys.begin(), keys.end(),
                   [](const std::string& s) { return tcpStore_.count(s) > 0; });
```

Reply:
Done